crates/cli/src/native_query/pipeline/match_stage.rs

```diff
@@ -41,61 +41,185 @@ fn check_match_doc_for_parameters_helper(
     input_document_type: &TypeConstraint,
     match_doc: Document,
 ) -> Result<()> {
-    if match_doc.keys().any(|key| key.starts_with("$")) {
-        analyze_document_with_match_operators(
-            context,
-            desired_object_type_name,
-            input_document_type,
-            match_doc,
-        )
-    } else {
-        analyze_document_with_field_name_keys(
-            context,
-            desired_object_type_name,
-            input_document_type,
-            match_doc,
-        )
+    for (key, value) in match_doc {
+        if key.starts_with("$") {
+            analyze_match_operator(
+                context,
+                desired_object_type_name,
+                input_document_type,
+                key,
+                value,
+            )?;
+        } else {
+            analyze_input_doc_field(
+                context,
+                desired_object_type_name,
+                input_document_type,
+                key,
+                value,
+            )?;
+        }
     }
+    Ok(())
 }
 
-fn analyze_document_with_field_name_keys(
+fn analyze_input_doc_field(
     context: &mut PipelineTypeContext<'_>,
     desired_object_type_name: &str,
     input_document_type: &TypeConstraint,
-    match_doc: Document,
+    field_name: String,
+    match_expression: Bson,
 ) -> Result<()> {
-    for (field_name, match_expression) in match_doc {
-        let field_type = TypeConstraint::FieldOf {
-            target_type: Box::new(input_document_type.clone()),
-            path: nonempty![field_name.into()],
-        };
-        analyze_match_expression(
-            context,
-            desired_object_type_name,
-            &field_type,
-            match_expression,
-        )?;
-    }
-    Ok(())
+    let field_type = TypeConstraint::FieldOf {
+        target_type: Box::new(input_document_type.clone()),
+        path: NonEmpty::from_vec(field_name.split(".").map(Into::into).collect())
+            .ok_or_else(|| Error::Other("object field reference is an empty string".to_string()))?,
+    };
+    analyze_match_expression(
+        context,
+        desired_object_type_name,
+        &field_type,
+        match_expression,
+    )
 }
 
-fn analyze_document_with_match_operators(
+fn analyze_match_operator(
     context: &mut PipelineTypeContext<'_>,
     desired_object_type_name: &str,
     field_type: &TypeConstraint,
-    match_doc: Document,
+    operator: String,
+    match_expression: Bson,
 ) -> Result<()> {
-    for (operator, match_expression) in match_doc {
-        match operator.as_ref() {
-            "$eq" => analyze_match_expression(
+    match operator.as_ref() {
+        "$and" | "$or" | "$nor" => {
```

Comment:
So once again we hardcode every function we want to support? This does at least result in very concrete checking code.